### DISCUSSION TITLE
What resources are being used to support the out of the box pico w

### DISCUSSION THREAD

**Discussion by microwinie** [earlephilhower/arduino-pico#1350] [General]
- Student Driver Alert - I'm looking at using a spi interface and I'd really like not collide with the wifi or console (serial/usb).  I'm using C/C++, if that matters, and would like to know what interfaces and pins I need to worry about, if any.  Any words of wisdom is appreciated.
btw, I find it slightly odd that all of the interfaces have set() functions for which pins to use however there is no get() function -perhaps git() could be a sanity check warning before mucking w/ any pins.

**Comment by earlephilhower** [comment]
As long as the pins you choose are legal for the HW and they don't overlap other peripherals, everything can be used at the same time.

**Reply by microwinie** [reply]
The kicker is the don't overlap others part and knowing where the are.

On Linux, I found the file ~/.arduino15/packages/rp2040/hardware/rp2040/3.1.0/pico-sdk/src/boards/include/boards/pico_w.h, which has some interface/pin definitions which (I hope) denotes what to stay away from.  I'll give this theory a test.
I've included a snippet from this file below.

Thank you for your reply.

// --- UART ---
#ifndef PICO_DEFAULT_UART
#define PICO_DEFAULT_UART 0
#endif
#ifndef PICO_DEFAULT_UART_TX_PIN
#define PICO_DEFAULT_UART_TX_PIN 0
#endif
#ifndef PICO_DEFAULT_UART_RX_PIN
#define PICO_DEFAULT_UART_RX_PIN 1
#endif


// --- I2C ---
#ifndef PICO_DEFAULT_I2C
#define PICO_DEFAULT_I2C 0
#endif
#ifndef PICO_DEFAULT_I2C_SDA_PIN
#define PICO_DEFAULT_I2C_SDA_PIN 4
#endif
#ifndef PICO_DEFAULT_I2C_SCL_PIN
#define PICO_DEFAULT_I2C_SCL_PIN 5
#endif

// --- SPI ---
#ifndef PICO_DEFAULT_SPI
#define PICO_DEFAULT_SPI 0
#endif
#ifndef PICO_DEFAULT_SPI_SCK_PIN
#define PICO_DEFAULT_SPI_SCK_PIN 18
#endif
#ifndef PICO_DEFAULT_SPI_TX_PIN
#define PICO_DEFAULT_SPI_TX_PIN 19
#endif
#ifndef PICO_DEFAULT_SPI_RX_PIN
#define PICO_DEFAULT_SPI_RX_PIN 16
#endif
#ifndef PICO_DEFAULT_SPI_CSN_PIN
#define PICO_DEFAULT_SPI_CSN_PIN 17
#endif

**Reply by earlephilhower** [reply]
That file is unused.  Check the variants/rpipico(w) headers instead.

But, again, if, say, SPI1.RX == GP5 but you don't actually use SPI1 then you can use GP5 as a regular GPIO.  Since you're the one building a board around the Pico only you know what's going to be where and what HW resources you're going to use.

**Comment by microwinie** [comment]
I just found your email and wanted to say 'thank you' for your reply to my post.

Btw, I made good progress with the pico/arduino SPI project.  The issue I had/have is that I've not seen a list of what resources are in use out of the box in terms of interfaces and pins.  Looking at the pico_w header file, I made (I hope) a decemt wag to use SPI1 and some GPIO pins AND both the wifi and usb serial lines still work :-)

I enjoy the ide, it's elegant for what it does and ease of use.

Thanks again.

________________________________
